flink-ml-lib/src/main/java/org/apache/flink/ml/feature/countvectorizer/CountVectorizer.java

```diff
@@ -0,0 +1,219 @@
+                        new VocabularyAggregator(getMinDF(), getMaxDF(), getVocabularySize()));
+
+        CountVectorizerModel model =
+                new CountVectorizerModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, getParamMap());
+        return model;
+    }
+
+    /**
+     * Extracts a vocabulary from input document collections and builds the {@link
+     * CountVectorizerModelData}.
+     */
+    private static class VocabularyAggregator
+            implements AggregateFunction<
+                    String[],
+                    Tuple2<Long, Map<String, Tuple2<Long, Long>>>,
+                    CountVectorizerModelData> {
+        private final double minDF;
+        private final double maxDF;
+        private final int vocabularySize;
+
+        public VocabularyAggregator(double minDF, double maxDF, int vocabularySize) {
+            this.minDF = minDF;
+            this.maxDF = maxDF;
+            this.vocabularySize = vocabularySize;
+        }
+
+        @Override
+        public Tuple2<Long, Map<String, Tuple2<Long, Long>>> createAccumulator() {
+            return Tuple2.of(0L, new HashMap<>());
+        }
+
+        @Override
+        public Tuple2<Long, Map<String, Tuple2<Long, Long>>> add(
+                String[] terms, Tuple2<Long, Map<String, Tuple2<Long, Long>>> vocabAccumulator) {
+            Map<String, Long> wc = new HashMap<>();
+            Arrays.stream(terms)
+                    .forEach(
+                            x -> {
+                                if (wc.containsKey(x)) {
+                                    wc.put(x, wc.get(x) + 1);
+                                } else {
+                                    wc.put(x, 1L);
+                                }
+                            });
+
+            Map<String, Tuple2<Long, Long>> counts = vocabAccumulator.f1;
+            wc.forEach(
+                    (w, c) -> {
+                        if (counts.containsKey(w)) {
+                            counts.get(w).f0 += c;
+                            counts.get(w).f1 += 1;
+                        } else {
+                            counts.put(w, Tuple2.of(c, 1L));
+                        }
+                    });
+            vocabAccumulator.f0 += 1;
+
+            return vocabAccumulator;
+        }
+
+        @Override
+        public CountVectorizerModelData getResult(
+                Tuple2<Long, Map<String, Tuple2<Long, Long>>> vocabAccumulator) {
+            Preconditions.checkState(vocabAccumulator.f0 > 0, "The training set is empty.");
+
+            boolean filteringRequired =
+                    !MIN_DF.defaultValue.equals(minDF) || !MAX_DF.defaultValue.equals(maxDF);
+            if (filteringRequired) {
+                long rowNum = vocabAccumulator.f0;
+                double actualMinDF = minDF >= 1.0 ? minDF : minDF * rowNum;
+                double actualMaxDF = maxDF >= 1.0 ? maxDF : maxDF * rowNum;
```

Comment:
It might be better to declare these two variables as integers, as they represent the number of documents.

Reply:
The benefit of using a double is that we don't need to care about whether to use `floor` or `ceil`, `>` or `>=` which makes the judgment more complex.

Reply:
Got it. I think the current implementation is OK then.